File: hw3/python/task3predict.py
import os
import sys
import json
import time


from pyspark import SparkContext


def predict_item(x, model_dict, business_index, user_index, user_business_dict, business_avg_rate, user_avg_rate, N):
    # x (user, business)
    user, business = user_index.get(x[0]), business_index.get(x[1])

    if business is None:
        p = user_avg_rate.get(user)
        return x, p
    if user is None or model_dict.get(business) is None:
        p = business_avg_rate.get(business)
        return x, p

    # if user_business_dict.get(user) is None or model_dict.get(business) is None:
    #     p = business_avg_rate.get(business)
    #     return x, p
    business_rate_dict = dict(user_business_dict[user])
    business_sim_dict = sorted(model_dict[business], key=lambda x: x[1], reverse=True)
    candidate_items = list()

    i = 0
    while len(candidate_items) < N and i < len(business_sim_dict):
        pair = business_sim_dict[i]
        if pair[0] in business_rate_dict.keys():
            candidate_items.append((pair[0], pair[1], business_rate_dict[pair[0]]))
        i += 1

    denominator = sum([abs(x[1]) for x in candidate_items])
    if denominator == 0:
        p = business_avg_rate.get(business) if business_avg_rate.get(business) is not None else user_avg_rate.get(user)
    else:
        p = sum([x[1] * x[2] for x in candidate_items]) / denominator

    return x, p


def predict_user(x, model_dict, business_index, user_index, business_user_dict, business_avg_rate, user_avg_rate, N):
    user, business = user_index.get(x[0]), business_index.get(x[1])
    r_u = user_avg_rate.get(user)

    # if user is None or business is None or business_user_dict.get(business) is None or model_dict.get(user) is None:
    #     p = user_avg_rate.get(user) if user_avg_rate.get(user) is not None else business_avg_rate.get(business)
    #     return x, p

    if user is None:
        p = business_avg_rate.get(business)
        return x, p
    if business is None or model_dict.get(user) is None:
        p = user_avg_rate.get(user)
        return x, p
    user_rate_dict = dict(business_user_dict.get(business))
    user_sim_dict = sorted(model_dict[user], key=lambda x: x[1], reverse=True)
    candidate_users = list()
    i = 0
    while len(candidate_users) < N and i < len(user_sim_dict):
        pair = user_sim_dict[i]
        if pair[0] in user_rate_dict.keys():
            candidate_users.append((pair[0], pair[1], user_rate_dict[pair[0]]))

        i += 1

    numerator = sum((x[2] - user_avg_rate[x[0]]) * x[1] for x in candidate_users)
    denominator = sum([abs(x[1]) for x in candidate_users])
    if denominator == 0:
        p = user_avg_rate[user] if user_avg_rate.get(user) is not None else business_avg_rate.get(business)
    else:
        p = r_u + numerator / denominator
    return x, p

if __name__ == '__main__':
    start_time = time.time()
    os.environ['JAVA_HOME'] = '/home/buyi/app/jdk'
    os.environ['PYSPARK_PYTHON'] = '/usr/bin/python3'
    # task3item.model task3item.predict item_based task3user.model task3user.predict user_based
    # train_file, test_file, model_file, output_file, cf_type = '/home/buyi/data/hw3/train_review.json', '/home/buyi/data/hw3/test_review.json', 'task3user.model', 'task3user.predict', "user_based"  # user_based
    train_file, test_file, model_file, output_file, cf_type = sys.argv[1:]
    sc = SparkContext('local[*]', 'task3')
    rdd_train = sc.textFile(train_file).map(json.loads)
    business_index = rdd_train.map(lambda x: x['business_id']).distinct().zipWithIndex().collectAsMap()
    user_index = rdd_train.map(lambda x: x['user_id']).distinct().zipWithIndex().collectAsMap()
    # business_index_dict = dict(rdd_train.map(lambda x: x['business_id']).distinct().zipWithIndex().collect())
    rdd_user = rdd_train.map(lambda x: (user_index[x['user_id']], x['stars'])).groupByKey().mapValues(lambda x: list(x))
    user_avg_rate = dict(rdd_user.map(lambda x: (x[0], sum(x[1]) / len(x[1]))).collect())
    rdd_business = rdd_train.map(lambda x: (business_index[x['business_id']], x['stars'])).groupByKey().mapValues(lambda x: list(x))
    business_avg_rate = dict(rdd_business.map(lambda x: (x[0], sum(x[1]) / len(x[1]))).collect())
    rdd_test = sc.textFile(test_file).map(json.loads).map(lambda x:(x["user_id"], x["business_id"])).filter(lambda x:x[0] in user_index.keys() or x[1] in business_index.keys())
    rdd_model = sc.textFile(model_file).map(json.loads)

    N = 5

    if cf_type == "item_based":
        user_business_dict = rdd_train.map(lambda x: (user_index[x['user_id']], (business_index[x['business_id']], x['stars']))).groupByKey().mapValues(lambda x: list(x)).collectAsMap()
        rdd_model_1 = rdd_model.map(lambda x: (business_index[x['b1']], business_index[x['b2']], x['sim'])).flatMap(lambda x: [(x[0], (x[1], x[2])), (x[1], (x[0], x[2]))]).groupByKey().mapValues(lambda x: list(x))
        model_dict = rdd_model_1.collectAsMap() # .filter(lambda x:x[1] in business_index.keys())
        result = rdd_test.map(
            lambda x: predict_item(x, model_dict, business_index, user_index, user_business_dict, business_avg_rate, user_avg_rate, N)).filter(lambda x:x[1] is not None).collect()

    else:
        business_user_dict = rdd_train.map(lambda x: (business_index[x['business_id']], (user_index[x['user_id']], x['stars']))).groupByKey().mapValues(
                lambda x: list(x)).collectAsMap()
        rdd_model_1 = rdd_model.map(lambda x: (user_index[x['u1']], user_index[x['u2']], x['sim'])).flatMap(
            lambda x: [(x[0], (x[1], x[2])), (x[1], (x[0], x[2]))]).groupByKey().mapValues(lambda x: list(x))

        model_dict = dict(rdd_model_1.collect())
        # print(model_dict) .filter(lambda x:x[0] in user_index.keys())
        result = rdd_test.map(
            lambda x: predict_user(x, model_dict, business_index, user_index, business_user_dict, business_avg_rate, user_avg_rate, N)).filter(lambda x:x[1] is not None).collect()
    with open(output_file, 'w') as f:
        for pair_rate in result:
            f.write(json.dumps(
                {"user_id": pair_rate[0][0], "business_id": pair_rate[0][1], "stars": pair_rate[1]}) + "\n")
    print("time: {}".format(time.time() - start_time))

File: hw3/scala/task3predict.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, PrintWriter}
import java.util.Date
import scala.math.abs

object task3predict {

  def predict_item(item: (String, String), model_dict: collection.Map[Long, List[(Long, Double)]], business_index: collection.Map[String, Long], user_index: collection.Map[String, Long], user_business_dict: collection.Map[Long, List[(Long, Double)]], business_avg_rate: collection.Map[Long, Double], user_avg_rate: collection.Map[Long, Double], N: Int): ((String, String), Double) = {
    var p = 0.0
    val u = user_index.get(item._1)
    val b = business_index.get(item._2)
    if (b.isEmpty) {
      if (user_avg_rate.contains(u.head)) {
        p = user_avg_rate.get(u.head).head
      }
      return (item, p)
    }
    if (u.isEmpty || !model_dict.contains(b.head)) {
      if (business_avg_rate.contains(b.head)) {
        p = business_avg_rate.get(b.head).head
      }

      return (item, p)
    }
    val user = u.head
    val business = b.head
    val business_rate_dict = user_business_dict.get(user).head.toMap
    val business_sim_dict = model_dict.get(business).head.sortBy(_._2).reverse
    var candidate_items:List[(Long, Double, Double)] = List()

    var i = 0
    while (candidate_items.length < N && i < business_sim_dict.length) {
      val pair = business_sim_dict.apply(i)
      if (business_rate_dict.keys.toList.contains(pair._1)) {
        candidate_items = candidate_items :+ (pair._1, pair._2, business_rate_dict.get(pair._1).head)
      }
      i += 1
    }

    val denominator = candidate_items.map(item=>abs(item._2)).sum

    if (denominator == 0) {

      if ( business_avg_rate.contains(business)) {
        p = business_avg_rate.get(business).head
      }
    }else {
      p = candidate_items.map(item=>item._2*item._3).sum / denominator
    }
    (item, p)
  }

  def predict_user(item: (String, String), model_dict: collection.Map[Long, List[(Long, Double)]], business_index: collection.Map[String, Long], user_index: collection.Map[String, Long], business_user_dict: collection.Map[Long, List[(Long, Double)]], business_avg_rate: collection.Map[Long, Double], user_avg_rate: collection.Map[Long, Double], N: Int): ((String, String), Double) = {
    var p = 0.0
    val u = user_index.get(item._1)
    val b = business_index.get(item._2)

    if (u.isEmpty) {
      if (business_avg_rate.contains(b.head)) {
        p = business_avg_rate.get(b.head).head
      }
      return (item, p)
    }
    if (b.isEmpty || !model_dict.contains(u.head)) {
      if (user_avg_rate.contains(u.head)) {
        p = user_avg_rate.get(u.head).head
      }

      return (item, p)
    }

    val business = b.head
    val user = u.head
    val r_u = user_avg_rate.get(user).head
    val user_rate_dict = business_user_dict.get(business).head.toMap
    val user_sim_dict = model_dict.get(user).head.sortBy(_._2).reverse
    var candidate_users:List[(Long, Double, Double)] = List()

    var i = 0
    while (candidate_users.length < N && i < user_sim_dict.length) {
      val pair = user_sim_dict.apply(i)
      if (user_rate_dict.keys.toList.contains(pair._1)) {
        candidate_users = candidate_users :+ (pair._1, pair._2, user_rate_dict.get(pair._1).head)
      }
      i += 1
    }

    val numerator = candidate_users.map(item=>(item._3 - user_avg_rate.get(item._1).head)*item._2).sum
    val denominator = candidate_users.map(item=>abs(item._2)).sum

    if (denominator == 0) {
      if ( user_avg_rate.contains(user)) {
        p = user_avg_rate.get(user).head
      }
    }else {
      p = r_u + numerator / denominator
    }
    (item, p)
  }

  def main(args: Array[String]) = {
    val start_time = new Date().getTime
    val conf = new SparkConf().setMaster("local[*]").setAppName("task1")
    val sc = new SparkContext(conf)
    //train_file, test_file, model_file, output_file, cf_type = '/home/buyi/data/hw3/train_review.json', '/home/buyi/data/hw3/test_review.json', 'task3user.model', 'task3user.predict', "user_based"  # user_based
    val train_file = args(0) //"E:\\Data\\inf553\\hw3\\publicdata\\train_review.json"
    val test_file = args(1) //"E:\\Data\\inf553\\hw3\\publicdata\\test_review.json"

    val model_file = args(2) //"task3item.scala.model" // task3user.model
    val output_file = args(3) //"task3item.scala.predict" // task3user.model

    val cf_type = args(4) //"item_based" //item_based


    val rdd_train = sc.textFile(train_file).map(item=>parse(item))
    val business_index = rdd_train.map(item=>compact(item \ "business_id")).distinct().zipWithIndex().collectAsMap()
    val user_index = rdd_train.map(item=>compact(item \ "user_id")).distinct().zipWithIndex().collectAsMap()

    val rdd_user = rdd_train.map(item=>(user_index.get(compact(item \ "user_id")).head, compact(item \ "stars").toDouble)).groupByKey().mapValues(_.toList)
    val user_avg_rate = rdd_user.map(item=>(item._1, item._2.sum/item._2.length)).collectAsMap()
    val rdd_business = rdd_train.map(item=>(business_index.get(compact(item \ "business_id")).head, compact(item \ "stars").toDouble)).groupByKey().mapValues(_.toList)
    val business_avg_rate = rdd_business.map(item=>(item._1, item._2.sum/item._2.length)).collectAsMap()

    val rdd_test = sc.textFile(test_file).map(item=>parse(item)).map(item=>(compact(item \ "user_id"), compact(item \ "business_id")))
      .filter(item=>user_index.keys.toList.contains(item._1) || business_index.keys.toList.contains(item._2))
    val rdd_model = sc.textFile(model_file).map(item=>parse(item))
    val N = 3
    if (cf_type == "item_based") {

      val user_business_dict = rdd_train.map(item=>(user_index.get(compact(item \ "user_id")).head, (business_index.get(compact(item \ "business_id")).head, compact(item \ "stars").toDouble)))
        .groupByKey().mapValues(_.toList).collectAsMap()
      val rdd_model_1 = rdd_model.map(item=>(business_index.get(compact(item \ "b1")).head,business_index.get(compact(item \ "b2")).head,compact(item \ "sim").toDouble))
        .flatMap(item=>{
          var l:List[(Long, (Long, Double))] = List()
          l = l :+ (item._1, (item._2, item._3))
          l = l :+ (item._2, (item._1, item._3))
          l
        }).groupByKey().mapValues(_.toList)
      val model_dict = rdd_model_1.collectAsMap()
      //.filter(item=>business_index.keys.toList.contains(item._2))
      val result = rdd_test
        .map(item=>predict_item(item, model_dict, business_index, user_index, user_business_dict, business_avg_rate, user_avg_rate, N)).filter(_._2 > 0)

      val pw = new PrintWriter(new File(output_file))
      result.collect().foreach(item=>pw.write("{\"user_id\": "+item._1._1+", \"business_id\": "+item._1._2+ ", \"sim\": " +item._2.toString + "}\n"))
      pw.close()
    }else {

      val business_user_dict = rdd_train.map(item=>(business_index.get(compact(item \ "business_id")).head, (user_index.get(compact(item \ "user_id")).head, compact(item \ "stars").toDouble)))
        .groupByKey().mapValues(_.toList).collectAsMap()
      val rdd_model_1 = rdd_model.map(item=>(user_index.get(compact(item \ "u1")).head,user_index.get(compact(item \ "u2")).head,compact(item \ "sim").toDouble))
        .flatMap(item=>{
          var l:List[(Long, (Long, Double))] = List()
          l = l :+ (item._1, (item._2, item._3))
          l = l :+ (item._2, (item._1, item._3))
          l
        }).groupByKey().mapValues(_.toList)

      val model_dict = rdd_model_1.collectAsMap()
      //.filter(item=>user_index.keys.toList.contains(item._1))
      val result = rdd_test
        .map(item=>predict_user(item, model_dict, business_index, user_index, business_user_dict, business_avg_rate, user_avg_rate, N)).filter(_._2 > 0)

      val pw = new PrintWriter(new File(output_file))
      result.collect().foreach(item=>pw.write("{\"user_id\": "+item._1._1+", \"business_id\": "+item._1._2+ ", \"sim\": " +item._2.toString + "}\n"))
      pw.close()
    }

    val end_time =new Date().getTime
    println("time: "+(end_time-start_time)/1000.0+"s")
  }
}
